File: Minsk/CodeAnalysis/Syntax/UnaryExpressionSyntax.cs
namespace Minsk.CodeAnalysis.Syntax;

public sealed class UnaryExpressionSyntax : ExpressionSyntax
{
    public SyntaxToken OperatorToken { get; }
    public ExpressionSyntax Operand { get; }

    public override SyntaxKind Kind => SyntaxKind.UnaryExpression;
    public override IEnumerable<SyntaxNode> Children => new SyntaxNode[] { OperatorToken, Operand };

    public UnaryExpressionSyntax(SyntaxToken operatorToken, ExpressionSyntax operand)
    {
        OperatorToken = operatorToken;
        Operand = operand;
    }
}

File: nim/src/minsk/codeAnalysis/syntax/expressions/unaryExpressionSyntax.nim
import minsk/codeAnalysis/syntax/[
  expressionSyntax,
  syntaxKind,
  syntaxNode,
  syntaxToken,
]

type
  UnaryExpressionSyntax* = ref object of ExpressionSyntax
    operatorToken*: SyntaxToken
    operand*: ExpressionSyntax

proc newUnaryExpressionSyntax*(
  operatorToken: SyntaxToken,
  operand: ExpressionSyntax
): UnaryExpressionSyntax =
  new(result)
  result.operatorToken = operatorToken
  result.operand = operand

method kind*(self: UnaryExpressionSyntax): SyntaxKind =
  return SyntaxKind.UnaryExpression

method children*(self: UnaryExpressionSyntax): seq[SyntaxNode] =
  result.add(self.operatorToken)
  result.add(self.operand)
